### DISCUSSION TITLE
Loading spinner

### DISCUSSION THREAD

**Discussion by emmamcbride** [UKHomeOffice/design-system#488] [Ideas]
**Loading screen/spinner** 

This is something that I've seen used in multiple Home Office services but not documented (to the best of my knowledge). These are used to help users understand that if they see this screen, they need to wait for something to happen in the background. 

For example: 
- This component is used in a case-working system when a user has finished one task and is automatically routed to complete a following task [while the task loads]. 
- This component was used in the designs of an identity verification web flow when the liveness scan had finished but the feedback was not ready.

<img width="868" alt="Screenshot 2023-01-17 at 15 11 44" src="https://user-images.githubusercontent.com/55453258/212935658-d0d4e2e6-8a5c-4995-8f61-82073d15b58f.png">

**Accessibility:** 
To make the spinner is accessible, ensure:

aria-live="polite"
role="status"
the div is called "loading"

When something goes wrong or loading times out, tell users there has been a problem and provide a link to another suitable screen. Do not leave the spinner loading indefinitely. The spinner does not use javascript. 

**HTML:**

```
 <div class="govuk-grid-row">
      <div class="govuk-!-padding-bottom-9"></div>
      <div class="govuk-!-padding-bottom-9"></div>
      <div class="loading" style="margin-left:auto; margin-right: auto" aria-live="polite" role="status"</div>
    </div>

    <div class="govuk-!-padding-bottom-7"></div>

    <div class="govuk-grid-row">
    <h1 class="govuk-heading-m" style="text-align: center">Loading the next task for this case</h1>
    <div class="govuk-!-padding-bottom-9"></div>
    <div class="govuk-!-padding-bottom-9"></div>
    </div>
```

**CSS:** 
The spinner is created using a CSS approach which should be smaller and easier to maintain.

 ```
 .loading {
  border: 12px solid #DEE0E2;
  border-radius: 50%;
  border-top-color: #005EA5;
  width: 80px;
  height: 80px;
  -webkit-animation: spin 2s linear infinite;
  animation: spin 2s linear infinite;
}
@-webkit-keyframes spin {
  0% { -webkit-transform: rotate(0deg); }
  100% { -webkit-transform: rotate(360deg); }
}
@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}
```

**Content:** 
Error message - tell the user there is something wrong with the next step. 
If the next task cannot be retrieved, times out or there is a system error, show this message to help the user return to the previous step, or do the next part of their journey. 

There is an existing GOV UK backlog discussion from 2018 - https://github.com/alphagov/govuk-design-system-backlog/issues/28

**Comment by guermanc** [comment]
Additional discussion that may be useful https://github.com/alphagov/govuk-design-system-backlog/issues/164

**Comment by tomyems** [comment]
Issue created to create loading spinner: https://github.com/UKHomeOffice/design-system/issues/494

**Comment by tomyems** [comment]
@emmamcbride there's a [question](https://github.com/UKHomeOffice/design-system/issues/494#issuecomment-1431531766) from @MikeDThorpe about the tags used for the loading text? Is it always `<h1>` tags? Is there anything else on the screen when there's a loading spinner? Might need to think about whether anything needs adding to the guidance ? (apologies if question is obvious)

**Comment by emmamcbride** [comment]
@tomyems That is a good question. I think in my opinion it should be the only thing on the screen as usually its waiting for the next screen to be ready but I will keep this in mind while I'm digging around to see how it's been used in other places. Thanks for highlighting

**Comment by tomyems** [comment]
Draft guidance and component: https://63f338575175e42dee239c77--hods.netlify.app/components?name=Loading%20spinner

**Comment by tomyems** [comment]
Updated component (with bold text - 'govuk-heading-m'): https://63f4d262acf7b52db2ea92dc--hods.netlify.app/components?name=Loading%20spinner

**Comment by tomyems** [comment]
Updated after content review: https://63fcab113b69be07fd6c00aa--hods.netlify.app/components?name=Loading%20spinner

**Comment by tomyems** [comment]
@emmamcbride We already have a blue (`#1d70b8`) used in the Home Office design system for status messages and timeline. What are your thoughts on using that blue instead of the one in the loading spinner (`#005EA5`)? Were there reasons behind your choice?

(A not great picture of the existing Home Office blue in the context of the loading spinner)
![image](https://user-images.githubusercontent.com/93599300/222673367-9653d3d7-bef1-4219-bd9e-d174fef1bf50.png)

**Comment by emmamcbride** [comment]
@tomyems From memory there's no reason why that other shade was used! Completely happy with using the existing colour and it makes more sense to - thanks for raising 👍

**Comment by joeformdesign** [comment]
Hello, thanks for the component and discussion. I'm working on a defra project (epr for packaging) where users need to upload a hefty csv file. We have a spinner on a page at the moment similar to this so might align with yours as the design seems more consistent with GDS. 

One question though, the guidance says the component does not require javascript. I understand the spinner does not require javascript but what the does spinner and page look like when javascript is disabled? Wouldn't the user be watching the spinner for years as there would be no refresh? Would we need to considered the interaction of no JS and get the user to refresh the page themselves?

**Comment by jsweet59** [comment]
Hello! I see this thread has been dormant for a while but my colleague came across this component and I had some queries/comments about it :) 

1. [Assistive technologies will generally only announce _dynamic_ changes in the content of a live region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Guides/Live_regions#:~:text=will%20generally%20only%20announce%20dynamic%20changes%20in%20the%20content%20of%20a%20live%20region) - so I don't think this component will actually announce anything via screenreader as there are no dynamic changes (for which you'd need JavaScript, contrary to component documentation). 
2. The loading spinner has no text value, so even if it did trigger a screenreader announcement there'd be nothing to announce. The h1 houses the message but is not programmatically linked in any way to the .loading div, so there's no reason to believe the h1 would be announced. 
3.  [Elements with the role status have an implicit aria-live value of polite](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/status_role#:~:text=Elements%20with%20the%20role%20status%20have%20an%20implicit%20aria%2Dlive%20value%20of%20polite%20and%20an%20implicit%20aria%2Datomic%20value%20of%20true) - so I don't think it's necessary to apply the aria-live attribute as well?

 ...so just interested to know if there's something I'm missing? 

Thanks in advance if anyone is still monitoring this thread!

**Reply by daniel-ac-martin** [reply]
Addresses in #650?

**Reply by guermanc** [reply]
Hi @jsweet59 thank you for taking the time to reach out! That all makes perfect sense and I've updated the guidance and implementation. Let me know if you have any further comments or suggestions.

**Comment by adamliptrot-oc** [comment]
Hi, is there a reason `prefers-reduced-motion` was not added to the CSS for this?
Despite its size such an animation would have a detrimental effect on users with certain conditions (for example vestibular). The animation is not essential to understanding the meaning so could be removed if the preference is detected.

**Comment by liammcmurray** [comment]
Hi! 👋 
We just used this in a prototype and it didn't test well.
Users associated it with the 'wheel of death' - something not working or not loading, especially as the results took around 10-15 seconds to be returned. and wanted more granular feedback.
We may use it conjunction with additional information for example ticking off a list of things happening in the background. It would be good to know if this approach has been done elsewhere.
